### ISSUE TITLE
How to properly use fidelity as a metric to evaluate different explanation models?

### ISSUE THREAD

**Issue by hegy1024** [AslanDing/Fidelity#1]
Hello, thanks for your exceptional work on `explainability evaluation`. Your papers present sufficient theoretical derivation to convince me that `robust fidelity` will be very helpful for my work. However, when I try to use `robust fidelity` on `BA2-Motifs` following the code in your file `tools/pyg_example.py` to **evaluate different explanation models**, I find that `fidelity` produces *little different results* for **explanation generated by GNNE** as well as for the **random explanation**. There is no significant distinction in numerical terms. I've tried other graph dataset, but I'm still getting similar results. I'm worried if I'm using your method in a bad way. So I want to know how to properly use `robust fidelity` to evaluate different explanation models. Any guidance or insights you could provide would be greatly appreciated.

**Comment by hegy1024** [comment]
Here are my result, it seems like that there are *little different results* for **explanation generated by GNNE** as well as for the **random explanation**: 
![1716302859587](https://github.com/AslanDing/Fidelity/assets/105419681/f6114c39-b7f3-4005-97af-17160110936c)
And here are my test code: 
```
path = osp.join(osp.dirname(osp.realpath(__file__)), '..', 'data', 'ba2')
dataset = BA2MotifDataset(path)
class GCN(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.conv1 = GCNConv(dataset.num_features, 20)
        self.relu1 = ReLU()
        self.conv2 = GCNConv(20, 20)
        self.relu2 = ReLU()
        self.conv3 = GCNConv(20, 20)
        self.relu3 = ReLU()
        self.lin = Linear(20 * 2, 2)

    def forward(self, x, edge_index, batch=None, edge_weights=None):
        if batch is None: # No batch given
            batch = torch.zeros(x.size(0), dtype=torch.long).to(x.device)
        embed = self.embedding(x, edge_index, edge_weights)

        out1 = global_max_pool(embed, batch)
        out2 = global_mean_pool(embed, batch)
        input_lin = torch.cat([out1, out2], dim=-1)

        out = self.lin(input_lin)
        return out

    def embedding(self, x, edge_index, edge_weights=None):
        if edge_weights is None:
            edge_weights = torch.ones(edge_index.size(1)).to(x.device)
        stack = []

        out1 = self.conv1(x, edge_index, edge_weights)
        out1 = torch.nn.functional.normalize(out1, p=2, dim=1)
        out1 = self.relu1(out1)
        stack.append(out1)

        out2 = self.conv2(out1, edge_index, edge_weights)
        out2 = torch.nn.functional.normalize(out2, p=2, dim=1)
        out2 = self.relu2(out2)
        stack.append(out2)

        out3 = self.conv3(out2, edge_index, edge_weights)
        out3 = torch.nn.functional.normalize(out3, p=2, dim=1)
        out3 = self.relu3(out3)

        input_lin = out3

        return input_lin


device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
model = GCN().to(device)
optimizer = torch.optim.Adam(model.parameters(), lr=0.01, weight_decay=5e-4)

model.load_state_dict(torch.load('./ba2_best_model')['model_state_dict'])

torch.manual_seed(2024)
torch.cuda.manual_seed(2024)
random.seed(2024)

explainer = Explainer(
    model=model,
    algorithm=GNNExplainer(epochs=200),
    explanation_type='model',
    node_mask_type='attributes',
    edge_mask_type='object',
    model_config=dict(
        mode='multiclass_classification',
        task_level='graph',
        return_type='log_probs',
    ),
)

target = 2
explanation = explainer(dataset[target].x.to(device), dataset[target].edge_index.to(device))
print(f'Generated explanations in {explanation.available_explanations}')

f_plus,f_minus = fidelity(explainer, explanation)

print("================= Explanation Edge Mask =======================")
# explanation
f_plus, f_minus, f_d, f_plus_label, f_minus_label,f_d_label = robust_fidelity(explainer,explanation,undirect=True,top_k=12)
print(f"f_plus: {f_plus:<10.4f}\n"
      f"f_minus: {f_minus:<10.4f}\n"
      f"f_d: {f_d:<10.4f}\n" 
      f"f_plus_label: {f_plus_label:<10.4f}\n"
      f"f_minus_label: {f_minus_label:<10.4f}\n"
      f"f_d_label: {f_d_label:<10.4f}\n")

# random explanation
explanation.edge_mask = torch.randn_like(explanation.edge_mask).sigmoid()
print("================= Random Edge Mask =======================")
f_plus,f_minus,f_d, f_plus_label,f_minus_label,f_d_label = robust_fidelity(explainer,explanation,undirect=True,top_k=12)
print(f"f_plus: {f_plus:<10.4f}\n"
      f"f_minus: {f_minus:<10.4f}\n"
      f"f_d: {f_d:<10.4f}\n" 
      f"f_plus_label: {f_plus_label:<10.4f}\n"
      f"f_minus_label: {f_minus_label:<10.4f}\n"
      f"f_d_label: {f_d_label:<10.4f}\n")
```

**Comment by flyingdoog** [comment]
Thanks for the question.  In robust fidelity, we have two important hyper-parameters, alpha1 and alpha2. By default, we set alpha1=1-alpha2=0.1.   In case the difference between comparison methods is marginal,  I suggest increasing  alpha1 and 1-alpha2.  For example, in a recent ICML 24 paper, https://www.vietanhnguyen.net/mage.pdf, they use alpha1=1-alpha2 = 0.2.

**Comment by hegy1024** [comment]
Thanks for responding to my question promptly and offering your suggestion. I had been using the default settings before.  After following your advice and modifying the hyper-parameters, I noticed some different outcomes compared to before. 
I will delve deeper into the correct usage of robust fidelity from the paper you shared. Thanks again for your help and for the excellent work you're doing.

**Comment by AslanDing** [comment]
Hi, sorry to reply so late.  In the example, I just use the default parameters. The difference is possible to be very small.  I will add some comments in the examples. @flyingdoog Thanks for the explanation!
